### PR TITLE
10. As a user, I want to be alerted when I’m entering an empty item or an item that's already on my list

### LINKED ISSUE #10: 10. As a user, I want to be alerted when I’m entering an empty item or an item that's already on my list
## Summary
Users shouldn't be able to add an empty item to their list, or add the same item twice. If they try to do this, we need to show them an error message that explains the problem. This way, we’ll prevent some clutter in their lists.

## Acceptance criteria
- [x] Show an error message if the user tries to submit an empty item
- [x] Show an error message if the user tries to submit a new item that is _identical_ to an existing name. For instance, if the list contains `apples` and the user adds `apples`.
- [x] Show an error message if the user tries to submit a new item that matches an existing name with punctuation and casing normalized. For instance, if the list contains `apples` and the user adds `aPples` or `apples,` or `a pples`.
- [x] The user’s original input is saved in the database

##  Notes
- There are a couple of ways to approach this task. If you’re not sure how to begin, _ask your mentors_ for advice!

### PR DESCRIPTION
**Pull request by StefieCaff**
## Description
Issue 10's purpose is to add error handling to the user input when adding a new list item to an existing list. 

- It does not accept empty item names including one or more spaces. 
- It does not allow the same word to be added regardless of case, punctuation, or white space, for example Frosted flakes, frosted fLakes, frosted flakes, fro sted flake s.
- It saves the item name entered by the user exactly.

The acceptance criteria does not address, and therefore permits the following instances:
- Users are allowed to add the plural form of an item that already exists (e.g., “apples” is allowed when “apple” is on the shopping list) and vice versa.
- Users are allowed to misspell the shopping item name.

## Related Issue
closes #10 

## Acceptance Criteria

- [x] Show an error message if the user tries to submit an empty item
- [x] Show an error message if the user tries to submit a new item that is _identical_ to an existing name. For instance, if the list contains `apples` and the user adds `apples`.
- [x] Show an error message if the user tries to submit a new item that matches an existing name with punctuation and casing normalized. For instance, if the list contains `apples` and the user adds `aPples` or `apples,` or `a pples`.
- [x] The user’s original input is saved in the database

## Type of Changes
`enhancement`

### Before
issue-10-before

### After
issue-10-after-2

issue10-after

## Testing Steps / QA Criteria
- git pull
- git checkout ec-ac-issue-10
- npm start
- navigate to home and choose a list
- if there are no items on the list, add an item
- attempt to add the same item testing case, white space, and punctuation
- navigate to firebase to check that the item was added exactly

### CODE REVIEW

**Review by hun-ah — COMMENTED** [review]
Looks great! Nice work :) I like the functions you created to check if the word exists/specific casing. 

Couple things: 
1. I notice that if I add whitespace before and after a word, the item will show up with at least one space in Firebase ie: " mango " or " mango" or "mango " when viewing in Firebase console. When I search for the item on the list item page, if I hit my space bar it will match items that have whitespace. For example I created an item "chair" with four spaces in front and when I hit my space bar four times it still matches the word because of the whitespace. But shows as " chair " in Firebase.

2. On the `manageList` page, I am seeing an error in the console about the input going from controlled to uncontrolled. That may have been there from before but I am noticing it now as I am testing your code. It must be coming from `addItem` on the input element.

Overall really nice work! The only thing that I'd consider before merging is the whitespace and if that's going to cause any problems.

**edit: I think the input error was something on my end because I'm not seeing it anymore!**

**Review by eonflower — APPROVED** [review]
Code seems clean and the functionality worked for me.

For future us to think about, having our list items render in a way where we could force the UI to capitalize the first letter of every word or something consistent. Just for a... prettier UI. But for now, this hits the spot!

**Comment by 3campos** [comment]
Thanks for the feedback and compliment on the functions, @hun-ah!

> 1. I notice that if I add whitespace before and after a word, the item will show up with at least one space in Firebase ie: " mango " or " mango" or "mango " when viewing in Firebase console.

Thanks for pointing this out. I had to click on the shopping list item (the actual string element) in Firestore to confirm that the white space reflects the user's input when multiple spaces are included at the start and end of the word. 

I investigated this, and for some arbitrary reason, the Firestore UI displays the shopping item name with a single space before or after the string representing our shopping item name when there is at least one space before and/or after the word. Firebase does this even when the string in fact contains multiple preceding or succeeding spaces inside the Firestore document. I originally thought that @StefieCaff and I were unintentionally trimming whitespace from words on the front end but Firebase was the culprit all along!

> 2. **edit: I think the input error was something on my end because I'm not seeing it anymore!**

Both @StefieCaff and I tried reproducing this error but could not. Since you aren't see it anymore, we will also disregard it for now.

**Comment by 3campos** [comment]
Thank you, @eonflower! I fully agree that we should consider doing something like that in the future as it will make the UI much prettier.

**Comment by hun-ah** [comment]
> I investigated this, and for some arbitrary reason, the Firestore UI displays the shopping item name with a single space before or after the string representing our shopping item name when there is at least one space before and/or after the word. Firebase does this even when the string in fact contains multiple preceding or succeeding spaces inside the Firestore document. I originally thought that @StefieCaff and I were unintentionally trimming whitespace from words on the front end but Firebase was the culprit all along!

Weird! It was just something to note. As long as you don't think it will cause any issues then everything is good!

**Review by hun-ah — APPROVED** [review]

File: src/components/AddItem.jsx [hunk 1]
```diff
@@ -1,20 +1,41 @@
...
 
+	const normalizedItemName = (str) => {
+		str
+			.toLowerCase()
+			.replace(/[^\w\s]|(\s+)/g, '')
+			.trim();
+	};
+
+	const existingItem = (itemName) => {
+		return data.some(
```
Comment (nick-zanetti) [inline src/components/AddItem.jsx:17]: Somehow I've never used this, you learn something new every day! 🙈

**Review by nick-zanetti — APPROVED** [review]
This looks good to me! Good job of being thorough with this and addressing even minor tweaks like the whitespace issue.

**Pull request merged**